lib/vector-buffers/src/topology/channel/sender.rs

```diff
@@ -259,23 +259,24 @@ impl<T: Bufferable> Sink<T> for BufferSender<T> {
                 // We need to block.  Nothing else to do, as the base sender will notify us when
                 // there's capacity to do the send.
                 WhenFull::Block => (Poll::Pending, SendState::Idle),
-                // We need to drop the next item.  We have to wait until the caller hands it over to
-                // us in order to drop it, though, so we pretend we're ready and mark ourselves to
-                // drop the next item when `start_send` is called.
-                //
-                // One "gotcha" here is that the base sender is still trying to reserve a sending
-                // slot for us, so technically it could complete between now and when we get to
-                // `start_send` and actually drop the item.
-                //
-                // Based on the current behavior of `PollSender<T>`, the best thing we can do here
-                // is to simply to drop the item and not abort the send, since that will leave
-                // `PollSender<T>` armed for the next time we call `poll_reserve`.  Since buffers
-                // are SPSC, there's no risk in trying up a sender slot.
-                //
-                // TODO: In the future, `PollSender<T>::start_send` may be tweaked to attempt a
-                // call to `Sender<T>::try_send` as a last ditch effort when `PollSender<T>` has not
-                // yet reserved the sending slot.  We could take advantage of this ourselves.
-                WhenFull::DropNewest => (Poll::Ready(Ok(())), SendState::DropNext),
+                WhenFull::DropNewest => {
+                    // If we tried to check for readiness, but got told that the base sender was not
+                    // yet ready, we need to also check to see if the base sender has capacity.  If
+                    // it does, then we've hit a situation where the Tokio cooperative scheduling
+                    // budget has been exceeded, and we're being asked to cooperatively yield to
+                    // allow other tasks to run.
+                    //
+                    // If we don't have capacity, though, then we should in fact drop this event.
+                    // One caveat here is that the disk buffers, aka the "opaque" senders, have no
+                    // concept of per-event capacity, so we'll always assume they are in fact at
+                    // their limit and we should drop.
```

Comment:
Well, so this comment only applies to disk v1, because we talk to `disk_v1::Writer` directly via its `Sink` interface.  As well, disk v1 never yields on the sender side, so from a correctness standpoint, the logic won't ever be wrong: it only returns `Poll::Pending` when the buffer size limit has been reached and it has to wait for the reader.

For disk v2, we actually use an interstitial channel to talk to an async task that performs the `Writer::write_record`/`Writer::flush` calls for each record... so disk v2 would also be subject to this issue, as well.  Technically speaking, the behavior is slightly different for disk v2 because the interstitial channel might legitimately fill up -- with or without this fix -- while the buffer itself is not truly at capacity, and thus report back that it's full and thus trigger the dropping behavior when configured with `DropNewest`.

Reply:
Aha, interesting!

For disk buffer v2, it seems undesirable for it to drop if just the intermediate channel is blocking, but the disk buffer itself isn't at capacity, no? That seems like it could happen simply if there is a spike in events that overwhelms the channel temporarily until the disk v2 channel reader can catch up? At the least, it seems to make the guarantees a bit weaker in a way we should probably document so users aren't confused if they see dropped events before the disk buffer is full.

Reply:
It's definitely suboptimal, yes, but the problem is that we have to serialize access to the writer, and doing it with a channel was the simplest possible initial implementation.

The spike PR I was working on to convert everything to `async fn` and remove `Sink` was heading towards the disk v2 writer being wrapped in an async mutex, which would allow the serialized access requirement to be met.

However, for both disk v1 and disk v2, I exposed fallible methods for attempting a write which essentially just went through the motions of doing a pre-flight "do I have capacity?" check, and if they didn't, they return early.  This should give us both the optimally correct behavior -- don't drop unless actually full -- without too much performance downside.

Reply:
As a side note, the vectorized event processing work should actually make things better, because there will be fewer individual sends into the channels overall, whether it's an in-memory buffer itself or the interstitial channel feeding the disk v2 buffer.

Reply:
I think given the timing and what my roadmap looks like with regards to getting the configuration schema work off the ground, 0.22 might be the most logical release for promoting disk v2 to stable... and in that context, I think the `async fn`/remove `Sink` work will have almost certainly landed.